File: src/client/acontext-py/src/acontext/resources/disks.py
"""
Disk and artifact endpoints.
"""

import json
from collections.abc import Mapping, MutableMapping
from typing import Any, BinaryIO, cast

from ..client_types import RequesterProtocol
from ..uploads import FileUpload, normalize_file_upload


def _bool_to_str(value: bool) -> str:
    return "true" if value else "false"


class DisksAPI:
    def __init__(self, requester: RequesterProtocol) -> None:
        self._requester = requester
        self.artifacts = DiskArtifactsAPI(requester)

    def list(self) -> Any:
        return self._requester.request("GET", "/disk")

    def create(self) -> Any:
        return self._requester.request("POST", "/disk")

    def delete(self, disk_id: str) -> None:
        self._requester.request("DELETE", f"/disk/{disk_id}")


class DiskArtifactsAPI:
    def __init__(self, requester: RequesterProtocol) -> None:
        self._requester = requester

    def upsert(
        self,
        disk_id: str,
        *,
        file: FileUpload
        | tuple[str, BinaryIO | bytes]
        | tuple[str, BinaryIO | bytes, str | None],
        file_path: str | None = None,
        meta: Mapping[str, Any] | MutableMapping[str, Any] | None = None,
    ) -> Any:
        upload = normalize_file_upload(file)
        files = {"file": upload.as_httpx()}
        form: dict[str, Any] = {}
        if file_path:
            form["file_path"] = file_path
        if meta is not None:
            form["meta"] = json.dumps(cast(Mapping[str, Any], meta))
        return self._requester.request(
            "POST",
            f"/disk/{disk_id}/artifact",
            data=form or None,
            files=files,
        )

    def get(
        self,
        disk_id: str,
        *,
        file_path: str,
        with_public_url: bool | None = None,
        with_content: bool | None = None,
        expire: int | None = None,
    ) -> Any:
        params: dict[str, Any] = {"file_path": file_path}
        if with_public_url is not None:
            params["with_public_url"] = _bool_to_str(with_public_url)
        if with_content is not None:
            params["with_content"] = _bool_to_str(with_content)
        if expire is not None:
            params["expire"] = expire
        return self._requester.request("GET", f"/disk/{disk_id}/artifact", params=params)

    def update(
        self,
        disk_id: str,
        *,
        file_path: str,
        meta: Mapping[str, Any] | MutableMapping[str, Any],
    ) -> Any:
        payload = {
            "file_path": file_path,
            "meta": json.dumps(cast(Mapping[str, Any], meta)),
        }
        return self._requester.request("PUT", f"/disk/{disk_id}/artifact", json_data=payload)

    def delete(
        self,
        disk_id: str,
        *,
        file_path: str,
    ) -> None:
        params = {"file_path": file_path}
        self._requester.request("DELETE", f"/disk/{disk_id}/artifact", params=params)

    def list(
        self,
        disk_id: str,
        *,
        path: str | None = None,
    ) -> Any:
        params: dict[str, Any] = {}
        if path is not None:
            params["path"] = path
        return self._requester.request("GET", f"/disk/{disk_id}/artifact/ls", params=params or None)

File: src/client/acontext-ts/src/resources/disks.ts
/**
 * Disk and artifact endpoints.
 */

import { RequesterProtocol } from '../client-types';
import { FileUpload, normalizeFileUpload } from '../uploads';
import { buildParams } from '../utils';
import {
  Artifact,
  ArtifactSchema,
  Disk,
  DiskSchema,
  GetArtifactResp,
  GetArtifactRespSchema,
  ListArtifactsResp,
  ListArtifactsRespSchema,
  ListDisksOutput,
  ListDisksOutputSchema,
  UpdateArtifactResp,
  UpdateArtifactRespSchema,
} from '../types';

export class DisksAPI {
  public artifacts: DiskArtifactsAPI;
  private requester: RequesterProtocol;

  constructor(requester: RequesterProtocol) {
    this.requester = requester;
    this.artifacts = new DiskArtifactsAPI(requester);
  }

  async list(options?: {
    limit?: number | null;
    cursor?: string | null;
    timeDesc?: boolean | null;
  }): Promise<ListDisksOutput> {
    const params = buildParams({
      limit: options?.limit ?? null,
      cursor: options?.cursor ?? null,
      time_desc: options?.timeDesc ?? null,
    });
    const data = await this.requester.request('GET', '/disk', {
      params: Object.keys(params).length > 0 ? params : undefined,
    });
    return ListDisksOutputSchema.parse(data);
  }

  async create(): Promise<Disk> {
    const data = await this.requester.request('POST', '/disk');
    return DiskSchema.parse(data);
  }

  async delete(diskId: string): Promise<void> {
    await this.requester.request('DELETE', `/disk/${diskId}`);
  }
}

export class DiskArtifactsAPI {
  constructor(private requester: RequesterProtocol) {}

  async upsert(
    diskId: string,
    options: {
      file:
        | FileUpload
        | [string, Buffer | NodeJS.ReadableStream]
        | [string, Buffer | NodeJS.ReadableStream, string | null];
      filePath?: string | null;
      meta?: Record<string, unknown> | null;
    }
  ): Promise<Artifact> {
    const upload = normalizeFileUpload(options.file);
    const files = {
      file: upload.asFormData(),
    };
    const form: Record<string, string> = {};
    if (options.filePath) {
      form.file_path = options.filePath;
    }
    if (options.meta !== undefined && options.meta !== null) {
      form.meta = JSON.stringify(options.meta);
    }
    const data = await this.requester.request('POST', `/disk/${diskId}/artifact`, {
      data: Object.keys(form).length > 0 ? form : undefined,
      files,
    });
    return ArtifactSchema.parse(data);
  }

  async get(
    diskId: string,
    options: {
      filePath: string;
      filename: string;
      withPublicUrl?: boolean | null;
      withContent?: boolean | null;
      expire?: number | null;
    }
  ): Promise<GetArtifactResp> {
    const fullPath = `${options.filePath.replace(/\/$/, '')}/${options.filename}`;
    const params = buildParams({
      file_path: fullPath,
      with_public_url: options.withPublicUrl ?? null,
      with_content: options.withContent ?? null,
      expire: options.expire ?? null,
    });
    const data = await this.requester.request('GET', `/disk/${diskId}/artifact`, {
      params,
    });
    return GetArtifactRespSchema.parse(data);
  }

  async update(
    diskId: string,
    options: {
      filePath: string;
      filename: string;
      meta: Record<string, unknown>;
    }
  ): Promise<UpdateArtifactResp> {
    const fullPath = `${options.filePath.replace(/\/$/, '')}/${options.filename}`;
    const payload = {
      file_path: fullPath,
      meta: JSON.stringify(options.meta),
    };
    const data = await this.requester.request('PUT', `/disk/${diskId}/artifact`, {
      jsonData: payload,
    });
    return UpdateArtifactRespSchema.parse(data);
  }

  async delete(
    diskId: string,
    options: {
      filePath: string;
      filename: string;
    }
  ): Promise<void> {
    const fullPath = `${options.filePath.replace(/\/$/, '')}/${options.filename}`;
    const params = { file_path: fullPath };
    await this.requester.request('DELETE', `/disk/${diskId}/artifact`, {
      params,
    });
  }

  async list(
    diskId: string,
    options?: {
      path?: string | null;
    }
  ): Promise<ListArtifactsResp> {
    const params: Record<string, string | number> = {};
    if (options?.path !== undefined && options?.path !== null) {
      params.path = options.path;
    }
    const data = await this.requester.request('GET', `/disk/${diskId}/artifact/ls`, {
      params: Object.keys(params).length > 0 ? params : undefined,
    });
    return ListArtifactsRespSchema.parse(data);
  }
}

